### PR TITLE
Feature to maximize widgets

### PR DESCRIPTION
**Pull request by willmcgugan**
Adds functionality to maximize widgets, which will show them full-screen in isolation.

- The screen has a `maximize` method which maximizes the given widget. It is possible to maximize anything, but it seems most appropriate for focused widgets.
- Widget subclasses can add a `ALLOW_MAXIMIZE` classvar to allow containers to be maximized. Handy for forms containing various inputs.
- Adds a system command to maximize widgets.

https://github.com/user-attachments/assets/77459dae-f5bc-4785-81a1-f72eae4d5295

### CODE REVIEW

File: src/textual/containers.py [hunk 1]
```diff
@@ -29,6 +29,10 @@ class Container(Widget):
 class ScrollableContainer(Widget, can_focus=True, inherit_bindings=False):
     """A scrollable container with vertical layout, and auto scrollbars on both axis."""
 
+    # We don't typically want to maximize scrollable containers,
+    # since the user can easily navigate the contents
+    ALLOW_MAXIMIZE = False
```
Comment (darrenburns) [inline src/textual/containers.py:34]: Not sure about this comment. The areas of Posting I've had people ask most for "maximise" in the past are scrollable. They just want to see more on screen at once.

File: src/textual/screen.py [hunk 2]
```diff
@@ -184,6 +186,11 @@ class Screen(Generic[ScreenResultType], Widget):
 
     Should be a set of [`command.Provider`][textual.command.Provider] classes.
     """
+    ALLOW_IN_MAXIMIZED_VIEW: ClassVar[str] = ".-textual-system,Footer"
```
Comment (darrenburns) [inline src/textual/screen.py:189]: Maybe?
```suggestion
    ALLOW_IN_MAXIMIZED_VIEW: ClassVar[str] = ".-textual-system,Header,Footer"
```

Comment (darrenburns) [inline src/textual/screen.py:189]: Does this only work with widgets on other layers and docked widgets?

File: src/textual/screen.py [hunk 3]
```diff
@@ -621,6 +674,41 @@ def focus_previous(self, selector: str | type[QueryType] = "*") -> Widget | None
         """
         return self._move_focus(-1, selector)
 
+    def maximize(self, widget: Widget, container: bool = True) -> None:
+        """Maximize a widget, so it fills the screen.
+
+        Args:
+            widget: Widget to maximize.
+            container: Maximize container if possible.
```
Comment (darrenburns) [inline src/textual/screen.py:682]: Could you add a bit more detail to the `container` part of this docstring?

File: src/textual/screen.py [hunk 4]
```diff
@@ -621,6 +674,41 @@ def focus_previous(self, selector: str | type[QueryType] = "*") -> Widget | None
...
+        """
+        if container:
+            # If we want to maximize the container, look up the dom to find a suitable widget
+            for maximize_widget in widget.ancestors:
```
Comment (darrenburns) [inline src/textual/screen.py:686]: If I pass a container into this method, won't this code mean it's going to maximise the first widget with `allow_maximize = True` instead of the container I passed in?

File: src/textual/screen.py [hunk 5]
```diff
@@ -621,6 +674,41 @@ def focus_previous(self, selector: str | type[QueryType] = "*") -> Widget | None
...
+                self.scroll_to_widget, self.focused, animate=False, center=True
+            )
+
+    def action_maximize(self) -> None:
+        """Action to maximize the currently focused widget."""
+        if self.focused is not None:
+            self.maximize(self.focused)
+
+    def action_minimize(self) -> None:
+        """Action to minimize the currently focused widget."""
```
Comment (darrenburns) [inline src/textual/screen.py:709]:
```suggestion
        """Action to minimize the currently maximized widget."""
```

File: tests/snapshot_tests/test_snapshots.py [hunk 6]
```diff
@@ -1488,3 +1489,41 @@ def test_scroll_page_down(snap_compare):
...
+
+        def compose(self) -> ComposeResult:
+            yield Button("Hello")
+            yield Button("World")
+            yield Footer()
+
+    assert snap_compare(MaximizeApp(), press=["m"])
+
+
+def test_maximize_container(snap_compare):
```
Comment (darrenburns) [inline tests/snapshot_tests/test_snapshots.py:1508]: I'd suggest you rename this to `test_maximize_widget_in_container` and then add another test that maximizes a container. From looking at the code I'm not 100% sure that's going to work as expected.

**Review by darrenburns — CHANGES_REQUESTED** [review]

File: src/textual/app.py [hunk 7]
```diff
@@ -3441,6 +3461,11 @@ async def _on_layout(self, message: messages.Layout) -> None:
         message.stop()
 
     async def _on_key(self, event: events.Key) -> None:
+        # Special case for maximized widgets
+        # If something is maximized, then escape should minimize
+        if self.screen.maximized is not None and event.key == "escape":
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/app.py:3466]: The widget should minimize.

File: src/textual/containers.py (hunk 1 above)
Comment (willmcgugan) [reply src/textual/containers.py:34]: But most containers we wouldn't want to be maximized. Apps tend to have a lot of containers, and we wouldn't want them to maximized for the most part. But you can always set that to `True` when you do.

File: src/textual/screen.py (hunk 2 above)
Comment (willmcgugan) [reply src/textual/screen.py:189]: The Header seemed less important. But the Footer contains stuff that may be required for the maximized widget.

I think its best to leave that to the dev.

Comment (willmcgugan) [reply src/textual/screen.py:189]: Only if they are a direct child of the Screen.

File: src/textual/screen.py (hunk 3 above)
Comment (willmcgugan) [reply src/textual/screen.py:682]: k

File: src/textual/screen.py (hunk 2 above)
Comment (darrenburns) [reply src/textual/screen.py:189]: In addition to what I said?

File: src/textual/screen.py (hunk 4 above)
Comment (willmcgugan) [reply src/textual/screen.py]: You don't pass in the container. You maximize a widget, and if that widget is within a maximizeable container, it will maximize that.

Feel free to suggest rewording to docstring to make that clearer.

File: tests/snapshot_tests/test_snapshots.py (hunk 6 above)
Comment (willmcgugan) [reply tests/snapshot_tests/test_snapshots.py:1508]: There are two tests, one that maximizes a widget, and one that maximizes a widget in a container. I think the covers the two cases?

File: src/textual/containers.py (hunk 1 above)
Comment (darrenburns) [reply src/textual/containers.py:34]: Isn't your comment here at odds with the current implementation though? Containers currently can be maximised if they're focusable?

File: src/textual/screen.py (hunk 2 above)
Comment (willmcgugan) [reply src/textual/screen.py:189]: Not sure what you are asking. Can you elaborate?

Comment (darrenburns) [reply src/textual/screen.py:189]: > Only if they are a direct child of the Screen.

Do you mean widgets on layers and docked widgets that are also direct children of the screen?

Or _any_ widget that is a direct child of the screen.

File: src/textual/screen.py [hunk 8]
```diff
@@ -621,6 +674,42 @@ def focus_previous(self, selector: str | type[QueryType] = "*") -> Widget | None
         """
         return self._move_focus(-1, selector)
 
+    def maximize(self, widget: Widget, container: bool = True) -> None:
+        """Maximize a widget, so it fills the screen.
+
+        Args:
+            widget: Widget to maximize.
+            container: If one of the widgets ancestors is a maximizeable widget, also maximize that.
```
Comment (darrenburns) [inline src/textual/screen.py:682]:
```suggestion
            container: If one of the widgets ancestors is a maximizeable widget, maximize that instead.
```

**Review by darrenburns — APPROVED** [review]
One suggestion for the docstring and one open question but LGTM.

File: src/textual/screen.py (hunk 2 above)
Comment (willmcgugan) [reply src/textual/screen.py:189]: Any widget. i.e. `Screen > *`

Comment (darrenburns) [reply src/textual/screen.py:189]: Got it, thanks 👍
